libs/execution_context/ebpf_link.c

```diff
@@ -314,25 +336,95 @@ static ebpf_result_t
+    return_value = _ebpf_link_instance_invoke_batch_begin(
+        extension_client_binding_context, sizeof(ebpf_execution_context_state_t), &state);
+
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    return_value = _ebpf_link_instance_invoke_batch(extension_client_binding_context, program_context, result, &state);
+    ebpf_assert(return_value == EBPF_SUCCESS);
+
+    return_value = _ebpf_link_instance_invoke_batch_end(extension_client_binding_context);
+    ebpf_assert(return_value == EBPF_SUCCESS);
+
+Done:
+    return return_value;
+}
+
+static ebpf_result_t
+_ebpf_link_instance_invoke_batch_begin(
+    _In_ const void* extension_client_binding_context, _In_ size_t state_size, _In_ void* state)
+{
+    bool enter_epoch = false;
+    bool provider_reference_held = false;
     ebpf_result_t return_value;
     ebpf_link_t* link = (ebpf_link_t*)ebpf_extension_get_client_context(extension_client_binding_context);
 
+    if (state_size < sizeof(ebpf_execution_context_state_t)) {
+        return_value = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
     if (link == NULL) {
         GUID npi_id = ebpf_extension_get_provider_guid(extension_client_binding_context);
         EBPF_LOG_MESSAGE_GUID(
             EBPF_TRACELOG_LEVEL_WARNING, EBPF_TRACELOG_KEYWORD_LINK, "Client context is null", npi_id);
         return_value = EBPF_FAILED;
-        goto Exit;
+        goto Done;
     }
 
     return_value = ebpf_epoch_enter();
-    if (return_value != EBPF_SUCCESS)
-        goto Exit;
-    ebpf_program_invoke(link->program, program_context, result);
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+    enter_epoch = true;
+
+    return_value = ebpf_program_reference_providers(link->program);
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+    provider_reference_held = true;
+
+    ebpf_get_execution_context_state((ebpf_execution_context_state_t*)state);
+
+Done:
+    if (return_value != EBPF_SUCCESS && provider_reference_held) {
+        ebpf_program_dereference_providers(link->program);
+    }
+
+    if (return_value != EBPF_SUCCESS && enter_epoch) {
+        ebpf_epoch_exit();
+    }
+
+    return return_value;
+}
+
+static ebpf_result_t
+_ebpf_link_instance_invoke_batch_end(_In_ const void* extension_client_binding_context)
+{
+    ebpf_link_t* link = (ebpf_link_t*)ebpf_extension_get_client_context(extension_client_binding_context);
+    ebpf_program_dereference_providers(link->program);
     ebpf_epoch_exit();
+    return EBPF_SUCCESS;
+}
+
+static ebpf_result_t
+_ebpf_link_instance_invoke_batch(
+    _In_ const void* extension_client_binding_context,
+    _Inout_ void* program_context,
+    _Out_ uint32_t* result,
+    _In_ const void* state)
+{
+    // No function entry exit traces as this is a high volume function.
+    ebpf_result_t return_value = EBPF_SUCCESS;
+    ebpf_link_t* link = (ebpf_link_t*)ebpf_extension_get_client_context(extension_client_binding_context);
```

Comment:
Given the point in line 422, should `ebpf_extension_get_client_context()` be moved to be an inline function in the .h file?

Reply:
I agree about removing the entry/exit tracing ebpf_extension_get_client_context. The decision on whether inline is more something that the PGO / optimizer should drive.

Reply:
Note: Because these are calls across translation unit boundaries, the C front-end compiler won't optimize this by default. The project does enable link time code generation, but without additional data the linker may or may not inline this based on its internal heuristic. The goal of the PGO is to provide feedback to LTCG etc. about which paths are hot should be inline / additionally optimized. 

Manually marking functions with the inline keyword has minimal impact when LTCG is enabled as it gives the linker much more flexibility in performing whole program optimizations such as seeing that this function is a conditional dereference of extension_client_binding_context and might be a good candidate for inlining.